vendor/go.etcd.io/etcd/client/v3/watch.go

```diff
@@ -688,6 +718,11 @@ func (w *watchGrpcStream) run() {
 	}
 }
 
+func shouldRetryWatch(cancelReasonError error) bool {
+	return (strings.Compare(cancelReasonError.Error(), v3rpc.ErrGRPCInvalidAuthToken.Error()) == 0) ||
+		(strings.Compare(cancelReasonError.Error(), v3rpc.ErrGRPCAuthOldRevision.Error()) == 0)
```

Comment:
this is stringifying the parameter error twice when it doesn't need to, and is stringifying constant API errors twice when they will never change

it's also unclear that exact string comparisons are correct for these... is there zero chance they will get wrapped by other errors that will add prefixes or suffixes? I'm surprised this isn't checking errors in a typed way

Reply:
cc @ahrtr @serathius

Reply:
> this is stringifying the parameter error twice


Indeed there is no need to stringify the error twice. I just delivered a PR to improve this: https://github.com/etcd-io/etcd/pull/14935

Thanks @liggitt for pointing this out.




> it's also unclear that exact string comparisons are correct for these.

It's correct here. etcd defines a set of [gRPC errors](https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/v3rpc/rpctypes/error.go#L24-L93) on server side. We need to translate each error to a string before responding to the client side. 

The logic here is that we need to retry on some error types (`ErrGRPCInvalidAuthToken` and `ErrGRPCAuthOldRevision`). So we need to compare the error message. Note that we have strict error messages definition, there is no any wrap & unwrap.

Reply:
> etcd defines a set of [gRPC errors](https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/v3rpc/rpctypes/error.go#L24-L93) on server side. We need to translate each error to a string before responding to the client side.

I'm really surprised the client is expected to compare error message strings to drive behavior. Isn't that fragile with skewed client and server versions?

Reply:
> I'm really surprised the client is expected to compare error message strings to drive behavior. Isn't that fragile with skewed client and server versions?

This is a valid concern. This should be a technical debt. We may consider to define client side [codes](https://github.com/grpc/grpc-go/blob/eeb9afa1f6b6388152955eeca8926e36ca94c768/codes/codes.go#L29) in future.

Reply:
This change is expanding the etcd API contract... the client now requires specific error strings from the server in order to behave correctly. Kubernetes describes working against a storage layer that implements the etcd wire protocol, but this change is expanding the etcd API contract in unstructured and undocumented ways.

I don't really think this is a great change in general, but I *really* don't think etcd should add this without explicit documentation of the exact message strings required to drive this behavior, and tests ensuring the server doesn't change in the future in ways that will break this.

cc @lavalamp @deads2k for visibility

Reply:
cc @mitake and @ptabor as well

Probably we should add a field something like `reasonCode` in [WatchResponse](https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/etcdserverpb/rpc.pb.go#L2192-L2221)



> Kubernetes describes working against a storage layer that implements the etcd wire protocol

Do you mean Kubernetes doesn't use the etcd `clientv3` at some places where it needs to talk to etcdserver at all?

Reply:
FYI. https://github.com/etcd-io/etcd/discussions/14992

Reply:
@dims Yes, will do it for 3.5 today.

Reply:
FYI. https://github.com/etcd-io/etcd/pull/14995